src/pubsub/blabber.rs

```diff
@@ -0,0 +1,172 @@
+
+                    sleep(Duration::from_millis(100)).await;
+                    continue;
+                }
+                Err(_) => {
+                    CONNECT_TIMEOUT.increment();
+                    PUBSUB_SUBSCRIBE_EX.increment();
+
+                    sleep(Duration::from_millis(100)).await;
+                    continue;
+                }
+            }
+        }
+
+        let mut s = stream.take().unwrap();
+
+        // read until response or timeout
+        loop {
+            match s.read(read_buffer.borrow_mut()).await {
+                Ok(n) => {
+                    unsafe {
+                        read_buffer.advance_mut(n);
+                    }
+                    {
+                        loop {
+                            let consumed = {
+                                let rbuf: &[u8] = read_buffer.borrow();
+
+                                if rbuf.len() >= HEADER_LEN as usize {
+                                    let len =
+                                        u32::from_be_bytes(rbuf[0..4].try_into().unwrap()) as usize;
+
+                                    // check if we have only a partial message
+                                    if rbuf.len() < len {
+                                        break;
+                                    }
+
+                                    let mut mbuf = rbuf[8..len].to_owned();
+
+                                    let _ = validator.validate(&mut mbuf);
+
+                                    // return the number of bytes consumed
+                                    len
+                                } else {
+                                    break;
+                                }
+                            };
+
+                            read_buffer.advance(consumed);
+                        }
+                    }
+                }
+                Err(_) => {
+                    PUBSUB_RECEIVE.increment();
+                    PUBSUB_RECEIVE_EX.increment();
+                    PUBSUB_SUBSCRIBER_CURR.sub(1);
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Launch tasks with one channel per task as gRPC is mux-enabled.
+pub fn launch_publishers(runtime: &mut Runtime, config: Config, work_receiver: Receiver<WorkItem>) {
+    debug!("launching blabber publisher tasks");
+
+    for _ in 0..config.pubsub().unwrap().publisher_poolsize() {
+        PUBSUB_PUBLISHER_CONNECT.increment();
+
+        // create one task per channel
+        for _ in 0..config.pubsub().unwrap().publisher_concurrency() {
+            runtime.spawn(publisher_task(work_receiver.clone()));
+        }
+    }
+}
+
+async fn publisher_task(work_receiver: Receiver<WorkItem>) -> Result<()> {
+    PUBSUB_PUBLISHER_CURR.add(1);
+
+    while RUNNING.load(Ordering::Relaxed) {
+        let _work_item = work_receiver
+            .recv()
+            .await
+            .map_err(|_| Error::new(ErrorKind::Other, "channel closed"))?;
+    }
+
```

Comment:
Don't we actually need to publish anything here? Or where is that actually happening?

Reply:
No. The server actively publishes without the client doing anything. I've simplified by removing all publisher tasks and adding a comment that hopefully makes this cleared going-forward.